src/interval.rs

```diff
@@ -53,28 +50,53 @@ impl Interval {
 impl<T: Oneshot> Interval<T> {
     ///Creates new instance with specified timer type.
     pub fn new(interval: time::Duration) -> Self {
-        Interval::Ongoing(T::new(interval), interval)
+        Self {
+            timer: T::new(interval),
+            interval,
+        }
+    }
+
+    #[inline(always)]
+    ///Stops interval
+    pub fn cancel(&mut self) {
+        self.timer.cancel()
+    }
+
+    ///Restarts interval
+    pub fn restart(&mut self, ctx: &task::Context) {
+        let interval = self.interval;
+        self.timer.restart(interval, ctx.waker());
+    }
+
+
+    #[inline(always)]
+    ///Gets mutable reference
+    pub fn as_mut(&mut self) -> &mut Self {
+        self
     }
 }
 
-impl<T: Oneshot> Future for Interval<T> {
-    type Output = Self;
+impl<T: Oneshot> Future for &'_ mut Interval<T> {
+    type Output = ();
 
     fn poll(mut self: Pin<&mut Self>, ctx: &mut task::Context) -> task::Poll<Self::Output> {
-        let mut state = Interval::Stopped;
-        mem::swap(self.as_mut().get_mut(), &mut state);
-        match state {
-            Interval::Ongoing(mut timer, interval) => match Future::poll(Pin::new(&mut timer), ctx) {
-                task::Poll::Ready(()) => {
-                    timer.restart(&interval, ctx.waker());
-                    task::Poll::Ready(Interval::Ongoing(timer, interval))
-                },
-                task::Poll::Pending => {
-                    *self = Interval::Ongoing(timer, interval);
-                    task::Poll::Pending
-                },
+        match Future::poll(Pin::new(&mut self.timer), ctx) {
+            task::Poll::Ready(()) => {
+                self.restart(ctx);
```

Comment:
I think in general the policy is for `Future` to never poll again `Ready` has been returned. This is the reason things like `FusedFuture` exist. Even though it makes for a nice API here, I think we may want to keep it the way it currently is for the `Future` impl (the `Stream` can be the way to do it with `&mut`)

Reply:
Yeah, but there is no requirements on `Future::poll` behavior so I want to keep this nice API to be able to do something like `interval.as_mut().await`